accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,495 @@
+
+# Detailed design
+
+The proposed solution is to make the `resolutions` field of the `package.json`
+file to be considered all the time and on a per-package basis (instead of
+only when the `--flat` parameter is used).
+
+When a nested dependency is being resolved by yarn, if the `resolutions` field
+contains a specification for this package, then it will be used instead.
+
+Special attention to the specific nature of package management in the npm
+ecosystem is given in this RFC: indeed, it is not unusual to have the same
+package being present as a nested dependency of multiple packages with
+different versions. It is thus possible within the `resolutions` field to
+express versions either for the whole dependency tree or only for a subset of
+it, using a syntax relying on glob patterns.
+
+Most of the examples are given with exact dependencies, but note that using a
+non-exact specification in the `resolutions` field should be accepted and
+resolved by yarn like it usually does. This subject is discussed below also.
+
+Any potentially counter-intuitive situation will result in a warning being
+issued. This subject is discussed at the end of this section.
+
+## Examples
+
+We have the following packages and their dependencies:
+```
+package-a@1.0.0
+ |_ package-d1@1.0.0
+     |_ package-d2@1.0.0
+
+package-a@2.0.0
+ |_ package-d1@2.0.0
+     |_ package-d2@1.0.0
+
+package-b@1.0.0
+ |_ package-d1@2.0.0
+     |_ package-d2@1.0.0
+
+package-c@1.0.0
+ |_ package-a@2.0.0
+     |_ package-d1@2.0.0
+         |_ package-d2@1.0.0
+```
+
+With:
+```json
+  "dependencies": {
+    "package-a": "1.0.0",
+    "package-b": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-d1": "2.0.0"
+  }
+```
+
+yarn will use `package-d1@2.0.0` for every nested dependency to `package-d1`
+and will behave as expected with respect to the `node_modules` folder by not
+duplicating the `package-d1` installation.
+
+With:
+```json
+  "dependencies": {
+    "package-a": "1.0.0",
+    "package-b": "1.0.0"
+  },
+  "resolutions": {
+    "package-a/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` only for `package-a` and `package-b` will
+still have `package-d1@2.0.0` in its own `node_modules`.
+
+With:
+```json
+  "dependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-a": "3.0.0"
+  }
+```
+
+`package-a` will still be resolved to `1.0.0`, but `package-c` will have
+`package-a@3.0.0` in its own `node_modules`.
+
+With:
+```json
+  "dependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "package-a": "3.0.0"
+  }
+```
+
+yarn will do nothing (see below why).
+
+With:
+```json
+  "dependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-a/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` both for `package-a` and the nested
+dependency `package-a` of `package-c`.

```

Comment:
Unless you mean that both `/package-a/package-d1` and `/package-c/package-a/package-d1` will be fixed at `3.0.0`?